tests/unit/s2n_protocol_version_getter_test.c

```diff
@@ -0,0 +1,395 @@
+            EXPECT_SUCCESS(s2n_stuffer_write_bytes(hello_stuffer, protocol_version, S2N_TLS_PROTOCOL_VERSION_LEN));
+            EXPECT_SUCCESS(s2n_stuffer_write(&server->handshake.io, &hello_stuffer->blob));
+
+            EXPECT_SUCCESS(s2n_client_hello_recv(server));
+
+            /* Ensure that a supported versions extension wasn't received. */
+            bool supported_versions_received = false;
+            EXPECT_SUCCESS(s2n_client_hello_has_extension(&server->client_hello, S2N_EXTENSION_SUPPORTED_VERSIONS,
+                    &supported_versions_received));
+            EXPECT_FALSE(supported_versions_received);
+
+            EXPECT_EQUAL(s2n_connection_get_server_protocol_version(server), server_version);
+            EXPECT_EQUAL(s2n_connection_get_client_protocol_version(server), client_hello_version);
+            EXPECT_EQUAL(s2n_connection_get_client_hello_version(server), client_hello_version);
+            EXPECT_EQUAL(s2n_connection_get_actual_protocol_version(server), client_hello_version);
+        }
+    }
+
+    /* Test protocol version getters on the client */
+    for (uint8_t server_version = S2N_SSLv3; server_version <= S2N_TLS13; server_version++) {
+        if (server_version == S2N_TLS13 && !s2n_is_tls13_fully_supported()) {
+            continue;
+        }
+
+        DEFER_CLEANUP(struct s2n_config *config = s2n_config_new(), s2n_config_ptr_free);
+        EXPECT_SUCCESS(s2n_config_add_cert_chain_and_key_to_store(config, chain_and_key));
+        EXPECT_SUCCESS(s2n_config_set_cipher_preferences(config, "test_all"));
+
+        DEFER_CLEANUP(struct s2n_connection *client = s2n_connection_new(S2N_CLIENT), s2n_connection_ptr_free);
+        EXPECT_NOT_NULL(client);
+        EXPECT_SUCCESS(s2n_connection_set_config(client, config));
+
+        DEFER_CLEANUP(struct s2n_connection *server = s2n_connection_new(S2N_SERVER), s2n_connection_ptr_free);
+        EXPECT_NOT_NULL(server);
+        EXPECT_SUCCESS(s2n_connection_set_config(server, config));
+        server->server_protocol_version = server_version;
+
+        DEFER_CLEANUP(struct s2n_test_io_pair io_pair = { 0 }, s2n_io_pair_close);
+        EXPECT_SUCCESS(s2n_io_pair_init_non_blocking(&io_pair));
+        EXPECT_SUCCESS(s2n_connection_set_io_pair(client, &io_pair));
+        EXPECT_SUCCESS(s2n_connection_set_io_pair(server, &io_pair));
+
+        EXPECT_OK(s2n_negotiate_test_server_and_client_until_message(server, client, SERVER_CERT));
+
+        EXPECT_EQUAL(s2n_connection_get_server_protocol_version(client), server_version);
+        EXPECT_EQUAL(s2n_connection_get_client_protocol_version(client), s2n_get_highest_fully_supported_tls_version());
+        EXPECT_EQUAL(s2n_connection_get_client_hello_version(client), S2N_TLS12);
+        EXPECT_EQUAL(s2n_connection_get_actual_protocol_version(client), server_version);
+    }
+
+    /* Test get_client_protocol_version fallback behavior on TLS 1.2 servers */
```

Comment:
I'm a little lost on what these last two tests are testing. What fallback are we talking about specifically?

Reply:
If a TLS 1.3 server receives a malformed supported versions extension (the extension is invalid or contains invalid supported versions), the connection fails. But TLS 1.2 servers ignore the extension, so the connection doesn't fail in this case, and applications may still call the client protocol version getter. So these tests ensure that the client protocol version assumed by the server (the client hello version) is reported instead.

I updated the comment - let me know if it's still unclear.

Reply:
For these tests in particular, why do we need a loop that checks both a valid extension and an invalid extension? Shouldn't the valid case already be checked by your earlier tests? 
I'm not trying to nit-picky, I just am having a hard time reading these tests so I'm looking for ways you can cut down on the complexity.

Reply:
Yeah the valid case is tested previously so it's not really needed here. I included this case just to prove that the test is setup correctly, in that the fallback really does occur due to the invalid extension and not because something with the test is weird. But I think this can be tested in a more straightforward way by checking to make sure that a supported versions extension was received, so I added that instead.

I also tried to further reduce the complexity of the tests by adding helper functions to write the supported versions extensions. Let me know if you see anything else that could be made more clear.